src/libraries/System.Private.CoreLib/src/System/SpanHelpers.T.cs

```diff
@@ -2,13 +2,160 @@
+                }
+                else
+                {
+                    T tmp = value; // Avoid taking address of the "value" argument. It would regress performance of the loops below.
+                    if (Unsafe.SizeOf<T>() == 1)
+                    {
+                        vector = new Vector<byte>(Unsafe.As<T, byte>(ref tmp));
+                    }
+                    else if (Unsafe.SizeOf<T>() == 2)
+                    {
+                        vector = (Vector<byte>)(new Vector<ushort>(Unsafe.As<T, ushort>(ref tmp)));
+                    }
+                    else if (Unsafe.SizeOf<T>() == 4)
+                    {
+                        vector = (Vector<byte>)(new Vector<uint>(Unsafe.As<T, uint>(ref tmp)));
+                    }
+                    else if (Unsafe.SizeOf<T>() == 8)
+                    {
+                        vector = (Vector<byte>)(new Vector<ulong>(Unsafe.As<T, ulong>(ref tmp)));
+                    }
+                    else
+                    {
+                        Debug.Fail("This should never happen. Did we miss special-casing a bitwise equatable type?");
+                        goto CannotVectorize;
+                    }
+                }
+
+                nuint totalByteLength = numElements * (nuint)Unsafe.SizeOf<T>(); // get this calculation ready ahead of time
+                nuint stopLoopAtOffset = totalByteLength & (nuint)(-Vector<byte>.Count * 2);
+                nuint offset = 0;
+
+                // Loop, writing 2 vectors at a time.
+
+                if (numElements >= 2 * (uint)(Vector<byte>.Count / Unsafe.SizeOf<T>()))
+                {
+                    do
+                    {
+                        Unsafe.WriteUnaligned(ref Unsafe.AddByteOffset(ref Unsafe.As<T, byte>(ref refData), offset), vector);
+                        Unsafe.WriteUnaligned(ref Unsafe.AddByteOffset(ref Unsafe.As<T, byte>(ref refData), offset + (uint)Vector<byte>.Count), vector);
+                        offset += 2 * (uint)Vector<byte>.Count;
+                    } while (offset < stopLoopAtOffset);
+                }
+
+                // There are [ 0, 2 * sizeof(Vector) ) bytes remaining.
+                // If there are >= sizeof(Vector) bytes remaining, write one vector now.
+
+                if ((totalByteLength & (nuint)Vector<byte>.Count) != 0)
+                {
+                    Unsafe.WriteUnaligned(ref Unsafe.AddByteOffset(ref Unsafe.As<T, byte>(ref refData), offset), vector);
+                }
+
+                // If there's any remaining space that won't fill a full vector, write a vector at
+                // the very end of the destination buffer. This will result in overwriting a previous
+                // entry, but that's ok since we're splatting the same value for all entries, so this
+                // won't result in data corruption.
+
+                if ((totalByteLength & (nuint)(Vector<byte>.Count - 1)) != 0)
```

Comment:
I believe you can omit this branch by just writing the vector at the end (idempotence, as you commented) always.

The check in L30 guarantees that at least one vector can be written. So just write it, and avoid this branch (with potential for mis-prediction).

The subtraction `totalByteLength - (uint)Vector<byte>.Count)` will be done at the address-resolution stage, so always writing should be cheaper than the branch.